src/common/libschedutil/ops.h

```diff
@@ -17,35 +17,38 @@
 extern "C" {
 #endif
 
-/* Callback for an alloc request.  jobspec is looked up as a convenience.
- * Decode msg with schedutil_alloc_request_decode().
- * 'msg' and 'jobspec' are only valid for hte duration of this call.
- * You should either respond to the request immediately (see alloc.h),
- * or cache this information for later response.
- */
-typedef void (schedutil_alloc_cb_f)(flux_t *h,
-                                    const flux_msg_t *msg,
-                                    const char *jobspec,
-                                    void *arg);
-
-/* Callback for a free request.  R is looked up as a convenience.
- * Decode msg with schedutil_free_request_decode().
- * 'msg' and 'R' are only valid for the duration of this call.
- * You should either respond to the request immediately (see free.h),
- * or cache this information for later response.
- */
-typedef void (schedutil_free_cb_f)(flux_t *h,
-                                   const flux_msg_t *msg,
-                                   const char *R,
-                                   void *arg);
-
-/* The job manager wants to cancel a pending alloc request.
- * The scheduler should look up the job in its queue.  If not found, do nothing.
- * If found, call schedutil_alloc_respond_cancel() and dequeue.
- */
-typedef void (schedutil_cancel_cb_f)(flux_t *h,
-                                     flux_jobid_t id,
-                                     void *arg);
+struct schedutil_ops {
+    /* Callback for an alloc request.  jobspec is looked up as a
+     * convenience.  Decode msg with schedutil_alloc_request_decode().
+     * 'msg' and 'jobspec' are only valid for hte duration of this
+     * call.  You should either respond to the request immediately
+     * (see alloc.h), or cache this information for later response.
+     */
+    void (*alloc)(flux_t *h,
+                  const flux_msg_t *msg,
+                  const char *jobspec,
+                  void *arg);
```

Comment:
Comment typo: "the".

Should we move `jobspec`  from `ops->alloc()` to `schedutil_alloc_request_decode()` now that it's included in `msg`?  (And be sure to note the fact that it is redacted of environment in that function's documentation).

Reply:
makes sense.  Perhaps same with `R` in `free`?  And add a `schedutil_cancel_request_decode()` and remove the job id from the cancel callback?  Looking through the history, I'm not entirely sure how some parameters were only in the callbacks but some in the decode functions.

Reply:
OR, alternately, `alloc_cb` and `free_cb` don't get passed a `flux_msg_t *msg`, they just get passed all the appropriate fields as parameters, no longer necessitating a `request_decode()` function?

Now that all of the operations are in the same struct, it seems weird that we're inconsistent with the style.

```
   int (*hello)(flux_t *h,                                                                                                                 
                 flux_jobid_t id,                                                                                                           
                 unsigned int priority,                                                                                                     
                 uint32_t userid,                                                                                                           
                 double t_submit,                                                                                                           
                 const char *R,                                                                                                             
                 void *arg);                                                                                                                
                                                                                                                                            
    void (*alloc)(flux_t *h,                                                                                                                
                  const flux_msg_t *msg,                                                                                                    
                  const char *jobspec,                                                                                                      
                  void *arg);                                                                                                               
                                                                                                                                            
    void (*free)(flux_t *h,                                                                                                                 
                 const flux_msg_t *msg,                                                                                                     
                 const char *R,                                                                                                             
                 void *arg);                                                                                                                
                                                                                                                                            
    void (*cancel)(flux_t *h,                                                                                                               
                   flux_jobid_t id,                                                                                                         
                   void *arg);            
```

two functions are passed `flux_msg_t *msg`, and two are just passed the fields directly.

Reply:
I think it's sort of handy to have the message.  Items that are parsed from it e.g. with `flux_msg_unpack()` don't need to be duplicated as long as the message exists.

`R` is not part of the message, so it can't be dropped  from the `free` callback.

I wonder actually the decode are all that useful now that we have an RFC for the message formats?  Maybe people would prefer to use `flux_msg_unpack()` on the messages.

Reply:
That would reduce the number of functions that we export in the API, and thus one less reason to "break" it. It would allow us to add fields to unpack without breaking existing users (not that I can think of a reason we'd do that)

Reply:
Ok, so I guess if we want consistency (which I think we should) we have to vote on a style of callback.

option 1, all parameters in the callback.  Pro: easiest for plugin writers, Con: any changes in protocol require ABI break
```
    void (*alloc)(flux_t *h,                                                                                                                
                  flux_jobid_t id,
                  uint32_t userid,
                  unsigned int priority,
                  double t_submit,
                  const char *jobspec,                                                                                                      
                  void *arg);  
```

option 2, no decode function, user parses everything from msg.  Pro: any changes to protocol won't break ABI, Con: annoying for plugin writers.  Bugs could permeate b/c plugin writers not aware of protocol change.

```
    void (*alloc)(flux_t *h,      
                  flux_msg_t *msg,                                                                                                          
                  void *arg);  
```

option 3, all params in decode function.  Basically this is the balance between the two above.  Protocol additions won't break ABI b/c can manually parse new additions from `msg`, may not break all priority plugins, etc.

```
    void (*alloc)(flux_t *h,      
                  flux_msg_t *msg,                                                                                                          
                  void *arg);  

    int schedutil_alloc_request_decode (const flux_msg_t *msg,
                                    flux_jobid_t *id,
                                    int *priority,
                                    uint32_t *userid,
                                    double *t_submit,
                                    const char **jobspec);
```

Since we're still in 0.X.Y version territory, my opinion is to go with option 1.  If we gotta change protocol, its annoying to break ABI, but that's why we're in alpha release territory.

Reply:
> won't break ABI b/c can manually parse new additions from msg, may not break all priority plugins,

er, s/priority plugins/schedulers.

I tend to like option 2 here and pointing scheduler authors to RFC 27.    If we are careful we can evolve that protocol without breaking schedulers, but if we do break the protocol, a failure to decode the message ought to be an obvious failure (scheduler interface stopping, with log 'protocol error' messages hopefully).

Reply:
Fwiw, I also prefer option 2, which more closely matches the design we've settled on for other interfaces in the rest of Flux.

Reply:
ok, option 2 it is.  (I had missed that you guys had made comments in between my 2 comments above, whoops)

Reply:
aww crud, missed a subtlety.

```
   int (*hello)(flux_t *h,                                                                                                                 
                 flux_jobid_t id,                                                                                                           
                 unsigned int priority,                                                                                                     
                 uint32_t userid,                                                                                                           
                 double t_submit,                                                                                                           
                 const char *R,                                                                                                             
                 void *arg);    
```

The hello protocol sends an array of entries and this callback is called per entry.  The `R` is looked up per entry as a convenience for the user.

I'm not too excited about having the user iterate through all entries themselves, although that'd probably be ok (they would have to iterate through entries in `priority` callback as well).

But I'm even less excited about them looking up `R` for each entry themselves.  At some point, `libschedutil` is barely even a utility library if we're having the user do everything related to these requests :P

Perhaps we should consider an option #4

```
    void (*alloc)(flux_t *h,
                  flux_msg_t *msg,                                                                                                             
                  flux_jobid_t id,
                  uint32_t userid,
                  unsigned int priority,
                  double t_submit,
                  const char *jobspec,                                                                                                      
                  void *arg);  
```

So we still remove the decode functions, since it's one less thing to support.  But we pass everything to the user in the callback that we know about now.  We also pass the `msg`, since that gives us some flexibility in the future to change the protocol and not break ABI.

Reply:
What I had envisioned for option 2 is that entities that are fetched from KVS as a convenience were still passed via the callback, but everything else available in the message is decoded with an unpack.

This would mean callbacks would have different signatures, but it is a good compromise between simplicity of the callback signatures vs extensibility of the interface.

Reply:
> What I had envisioned for option 2 is that entities that are fetched from KVS as a convenience were still passed via the callback, but everything else available in the message is decoded with an unpack.
> 
> This would mean callbacks would have different signatures, but it is a good compromise between simplicity of the callback signatures vs extensibility of the interface.

Generally speaking agree, but found it difficult to handle the `R` case for the `hello_cb`.  If we stick with the option 2 plan, what would you think works best for passing `R` to the callback?

One idea:

```
   int (*hello)(flux_t *h, 
                 flux_msg_t *msg,
                 int index,
                 const char *R,                                                                                                             
                 void *arg);    
```

so the caller knows which index in the array of jobs the R corresponds to?

then if we do the above, we might want to do the same for the `prioritize` cb as well (even though we aren't passing `R` to it).

Or we could pass an array of `R` objects to the callback?

```
   int (*hello)(flux_t *h, 
                 flux_msg_t *msg,
                 const char **R,
                 void *arg);    
```

neither of these seemed too inviting to me.  But perhaps I'm not creative enough to think of a good solution :-)

Reply:
I may not be remembering correctly, but doesn't the current callback get called once per job? Does this need to change? Then the hello typedef is simply
```c
    int (*hello_f) (flux_t *h, const flux_msg_t *msg, const char *R, void *arg)
```

Reply:
> I may not be remembering correctly, but doesn't the current callback get called once per job?

Yes.

>  int (*hello_f) (flux_t *h, const flux_msg_t *msg, const char *R, void *arg)

But if the message contains information about multiple jobs, how will the user know which job `R` belongs to?

Reply:
For the `hello` callback `flux_msg_unpack (msg, "{s:I}", "id", &jobid);` would return the jobid. That is, we might need to create a fake msg for each job for the multiple-jobs in a message case. In fact, if we did this then maybe `R` could be put in the fake message as well. If this rankles, we could introduce some other type (similar to `flux_plugin_arg_t`) through which these args are passed.

For the `sched.prioritize` RPC, I would just document the message format and have the callback itself iterate of the jobids instead of trying to make a callback per job. This would allow the scheduler to re-sort its queue once after each message, instead of each job.

Reply:
> That is, we might need to create a fake msg for each job for the multiple-jobs in a message case.  In fact, if we did this then maybe R could be put in the fake message as well.

Ahh, I didn't know that's what where you were going with this.  I dunno, it feels weird to me that we would say "follow RFC27 for parsing messages, except for the hello protocol case".  If we're making a special exception for the hello protocol, no reason we can't just keep it the way it is?

>  If this rankles, we could introduce some other type (similar to flux_plugin_arg_t) through which these args are passed.

I like this idea.  It would solve everything we've been talking about.  Not-breaking ABI on changes, same callback functions, etc.  Of course, it is a lot to add/change.

I will admit I'm a bit biased towards the "be consistent" side of things, but that's my personality.  I will go with how everyone else feels/votes on this.  Shall hello callback just be different than the rest?  If so, should we add `R` into the fake msg?

Reply:
I suppose one possibility is we could alter [RFC 27](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_27.html) to make `job-manager.hello` a multi-response RPC.  Then we could include RFC-defined message plus R in the callback and be somewhat consistent with the alloc callback?

Reply:
Just noticed that fluxion ignores `R` in the free callback.  Should we introduce flags somewhere that customize what's available in the callback?

Reply:
> Should we introduce flags somewhere that customize what's available in the callback?

That seems like a good idea if we're otherwise doing a kvs lookup for nothing.

Reply:
> I suppose one possibility is we could alter RFC 27 to make job-manager.hello a multi-response RPC. Then we could include RFC-defined message plus R in the callback and be somewhat consistent with the alloc callback?

I had considered that too, but resisted the call to change the RFC to match API desires.  Wouldn't there be performance implications sending a million RPCs, each only containing info about 1 job?

Reply:
> > Should we introduce flags somewhere that customize what's available in the callback?
> That seems like a good idea if we're otherwise doing a kvs lookup for nothing.

I like the idea too.  Created issue #3465 to keep track (don't want to lose track of all `libschedutil` desired changes).

Reply:
> I had considered that too, but resisted the call to change the RFC to match API desires. Wouldn't there be performance implications sending a million RPCs, each only containing info about 1 job?

That doesn't seem too bad to me - it's only _running_ jobs, and only at restart.

Reply:
submitted https://github.com/flux-framework/rfc/pull/283 as potential RFC27 change